(Design) Document the open questions around standalone markup (#542)
unicode-org/message-format-wg: exploration/open-close-expressions.md

* Introduces two new sigils, the pound sign `#` and the forward slash `/`.

> 

Reviewer: I would not say "elements" (use placeholder instead). HTML has elements, but we don't.

A different way of saying this is:

```suggestion
There is no difference between an "open" and a "standalone" placeholder in terms of syntax.
It is up to the implementation as to whether a given placeholder is treated
as an opening item or an isolated item in the output.
```

> 

Reviewer 2: Updated my suggestions above with `s/elements/placeholders/`.

> 

Author: I disagree with this suggestion, at least until we discuss the open questions related to standalone elements. The whole point of this PR is to emphasize that the current proposal is incomplete without our answering these questions.

Note that the "hash and slash" proposal did include a way to express standalone markup, and so did the "plus and minus" proposal. I'm concerned that suddenly we're discussing a different proposal in which standalone markup has been removed.

How about we add the following con instead of the ones you both proposed?

```suggestion
- Doesn't answer how to handle standalone markup; see below.
```

> 

Reviewer: @stasm I don't think that's quite correct. The proposal does say how to handle standalone markup, but not in a syntactically distinct way.

I think there is a case to be made for separate standalone syntax, as it allows relatively naive implementations to be created that don't have a list of standalone vs. spanning markup. 

Thus I think your "con" would be better phrased as:

```suggestion
- Standalone placeholders are not syntactically distinct from opening placeholders; 
  implementations require out-of-band information to distinguish standalone 
  from opening placeholders.
```

> 

Reviewer 3: I hope I don't seem contrary but £ is called a pound sign. # is called an octothorpe.
Coming from a country that used to have pounds as a unit of currency.

> 

Reviewer: @kenguest In the specification it would be _U+0023 NUMBER SIGN `#`_ because that is its Unicode name. Octothorpe and pound sign are both alternate names 😉

> 

Author: > The proposal does say how to handle standalone markup, but not in a syntactically distinct way.

But it’s not *the same* proposal that was called “hash and slash” which was added in #517.

I object how #535 removed the standalone syntax from that proposal, quoting consensus which was not officially discussed.

> 

Reviewer 2: I remain somewhat confused by the desire to remove this documented "con" of the currently proposed design. This PR and our discussions around this issue have made it clear that there's still desire to discuss standalone markup, but the proposed design itself is not changed here. It still does not differentiate in syntax standalone and open placeholders. Why can't we mention that? Either with the original text, or that proposed by @aphillips or myself earlier in this thread.

On a somewhat meta level, in reply to @stasm from https://github.com/unicode-org/message-format-wg/pull/542#discussion_r1413209532:
> The whole point of this PR is to emphasize that the current proposal is incomplete without our answering these questions.

This approach makes this seem like this PR is attempting re-re-define how we do design and how we work with design documents. Thus far, we've been able to iterate on these docs and their proposed solutions while keeping them internally coherent. Witness for instance the changes to this document and its proposed design. Why can't we continue doing so?

We clearly intend to discuss standalone markup further; I don't see how leaving this out helps that. Or are you suggesting that "standalone" and "open" using the same syntax is _not_ a disadvantage of the proposed design?

> 

Author: > I remain somewhat confused by the desire to remove this documented "con" of the currently proposed design. This PR and our discussions around this issue have made it clear that there's still desire to discuss standalone markup, but the proposed design itself is not changed here. It still does not differentiate in syntax standalone and open placeholders. 

It used to, however, right? And then #535 changed it, and I’m trying to object how it did it. See my comment in that PR.

I appreciate the attempt to limit the scope of the proposal to just open and close, but I also note that by doing so, we effectively make a decision about standalone.

I’d prefer it if the doc included `{#foo/}`, as it used to previously. But I also acknowledge that we will discuss this further on Monday.

Since I’m away this week, I’ll defer to you and @aphillips. And since I’m typing this on my phone, would you mind taking over so that we’re able to merge this PR soon? Thanks!

* Introduces two new sigils, the pound sign `#` and the forward slash `/`.

* In Mustache, the `{{#foo}}`...`{{/foo}}` syntax is used for *control flow* statements rather than printable data.

### Standalone Markup

_Standalone_ markup remains an open question.
Three alternatives have been suggested:

* Introduce additional syntax inspired by XML: `{#foo/}`.

>   This approach encodes the _standalone_ aspect in the data model, and doesn't rely on external sources of truth,

Reviewer: Not just the data model: it's in the syntax too!

Author: This is mentioned in the same sentence, which continues onto the next line: `at the expense of introducing new syntax, which is admittedly a bit clunky.`

  at the expense of introducing new syntax, which is admittedly a bit clunky.

  It also creates a dichotomy between standalone _markup_ and regular _placeholders_, which can be argued to be _standalone_ as well.

* Differentiate _open_ and _standalone_ elements based on the information stored in the registry,
  or based on some other external source of truth
  (e.g. the structure of the source to which the message is applied).

  This approach offers a cleaner syntax at the expense of relying on external sources of truth for telling open and standalone elements apart.

>   Without the access to these sources of truth (e.g. a custom registry), standalone markup will be interpreted as open elements.

Reviewer: Without access to some implementation, _all_ forms of markup are just so much noise.

I'm slightly sad that we didn't show using namespaces in the design document, since I suspect that `{#html:span id=foo}I am spanned{/html:span}` is friendlier to tools, given `{#ttml:span id=foo}I am timed text{/ttml:span}` exists too. Namespacing suggests how implementations would solve the problem of "external knowledge". There's no way that any of this works without some form of external information anyway.

Author: Yes, but you can get a lot of mileage without an implementation, just working with the data model, especially in tooling.

Reviewer: I think namespaces are important for markup regimes to work. They won't be built-in to the default registry, so there needs to be a way to plug them in and for tools _and implementations_ to know what they mean.

* Re-use regular placeholders with regular function calls: `{:img src=logo.png}`.

  This approach leverages the fact that regular placeholders can also be thought of as being _standalone_.
  It also doesn't introduce any new syntax, instead relying on the function call syntax: `:img`.

  Namespacing is a concern: to ensure forward compatibility, implementations and users should avoid introducing new symbols to the default, anonymous namespace.
  Thus, a custom function representing a standalone element should be namespaced: `:html:img` or `:moz:img`.
  While a good practice in general, it results in more verbose syntax,
  possibly also for the _open_ and _close_ markup, which would be namespaced for consistency: `{#html:link}...{/html:link}`.

>   In this approach there's also no way to enforce the lack of operands to standalone markup at the syntax level.
>   Instead, access to a custom registry is required to know that `:img` does not accept any operands.

Reviewer: I would say this differently.

```suggestion
In this approach, there's no way to distinguish standalone markup for regular functions.
This also means that there is no way to enforce a lack of operands for standalone
(as we do for open/close).
The open/close placeholders are default ignorable in certain formatting operations
because they can be distinguished from normal placeholders.
With `:img`, the placeholder would appear in all formatting operations, possibly as the
replacement "logo" when unsupported.
```
